### ISSUE TITLE
Filter balk zoals bij wildschade

### ISSUE THREAD

**Issue by SanderDevisscher** [inbo/reporting-rshiny-grofwildjacht#148]
In plaats van per grafiek (#147, #146, #145 & #144) dezelfde keuzes te moeten maken is het misschien nuttig een filter balk zoals bovenaan wildschade te voorzien bij de exoten pagina.

deze zou dan de volgende selectie velden bevatten:

- [ ]  periode (slider)
- [ ]  taxonomy (kingdom, phylum, ... t.e.m. family) 
Deze velden gedragen zich als filter voor de lagere hierachische velden.
Kingdom >filtert> phylum > filtert> ...

- [ ]  locality (België (BE), Vlaanderen (BE-VLG), Brussel (BE-BRU), Wallonië (BE-WAL)) => BE als default
- [ ]  degree of establishment
- [ ]  pathway_level 1 & pathway_level 2
- [ ]  habitat
- [ ]  source

Deze selectievelden starten dan steeds met een default waarde (locality) of alle opties geselecteerd

**Comment by mvarewyck** [comment]
@timadriaens @damianooldoni 

GENERAL
- Where can we find the R-code that was used to make the existing plots?
E.g. 
![exoten](https://user-images.githubusercontent.com/22143743/77404510-232d5080-6db2-11ea-9861-7df6235fba59.png)
- Filter balk: Will the user always filter on all these variables?
- For which of these filters can user select single choice versus multiple choices (I suppose source, maybe also degree of est.)?
- What are the default selected choices?

DATA
- periode: corresponds to which variable? first_observed, last_observed? I found these variables have respectively range from 1201 and 1732 onwards, while in the graphs the year starts at 1920. Extra filtering of data needed beforehand?
- degree of establishment: These are the values that I detect in the data, can you provide labels for all of them, e.g. the 3rd, blackburn_et_al_2011:E?
```
> unique(exotenData[, "degree of establishment"])
            degree of establishment
 1:                                
 2: established (blackburn_2011:C3)
 3:          blackburn_et_al_2011:E
 4:         blackburn_et_al_2011:C0
 5:         blackburn_et_al_2011:C3
 6:         blackburn_et_al_2011:B2
 7:      casual (blackburn_2011:C1)
 8:    released (blackburn_2011:B3)
 9:  colonizing (blackburn_2011:D1)
10: reproducing (blackburn_2011:C2)
11:     captive (blackburn_2011:B1)
12:         blackburn_et_al_2011:D2
```
- source: I've detected 2838 different values for that. Is it really the purpose that the user selects one or more of these? Should we provide option to select all sources?

MISSING VALUES
```
> sapply(exotenData, function(x) sum(is.na(x)))
         first_observed           last_observed                 kingdom 
                   6401                    6401                       0 
                 phylum                   class                   order 
                      0                       0                       0 
                 family                locality            native range 
                      0                       0                       0 
degree of establishment          pathway_level1          pathway_level2 
                      0                       0                       0 
                 marine              freshwater             terrestrial 
                    154                     154                     154 
                 source 
                      0 
> nrow(exotenData)
[1] 14262
```
You can see that the observation date is missing in almost 50% of the cases. This would mean that all these records are excluded from the graphs that display "periode" or filter on "periode".

**Comment by mvarewyck** [comment]
> Where can we find the R-code that was used to make the existing plots?

Found the source: e.g. https://github.com/trias-project/indicators/blob/master/src/indicator_introductions_per_year.Rmd

**Comment by SanderDevisscher** [comment]
> Filter balk: Will the user always filter on all these variables?

Yes, I see this as a filter like at the top of the wildschade page

> For which of these filters can user select single choice versus multiple choices (I suppose source, maybe also degree of est.)?

Here is what I suggest. @timadriaens what do you think 
- periode (slider) 
   - **Default = 1950 - now**
   - **Slider based on first_observed** 
- taxonomy (kingdom, phylum, ... t.e.m. family) 
   - **Default = empty**
   - **Choice multiple**
   - Lower level filter only display when higher level filter != empty
      - levels: kingdom -> phylum -> class -> order -> family 
   - Choices in kingdom filters options in phylum
   - Choices in phylum filters options in class
   - Choices in class filters options in order
   - Choices in order filters options in family
-  locality (België (BE), Vlaanderen (BE-VLG), Brussel (BE-BRU), Wallonië (BE-WAL)) 
     - **Default = empty (== België)**
     - **Multiple choice**
     - New choices should overwrite België filter
-  degree of establishment 
    - **Default =empty**
    - **Multiple choice** 
-  pathway_level 1 & pathway_level 2
   - **Default =empty**
   - **multiple choice**
   - choice in pathway_level 1 filters options in pathway_level 2
   - pathway_level 2 filter only displays when pathway_level 1 filter  != empty
-  habitat
   - **Default =empty**
   - **multiple choice**
-  source
   - **Default =empty**
   - **multiple choice**

> periode: corresponds to which variable? first_observed, last_observed? I found these variables have respectively range from 1201 and 1732 onwards, while in the graphs the year starts at 1920. Extra filtering of data needed beforehand?

you should use first_observed. filtering should be done according to the period slider.

> degree of establishment: These are the values that I detect in the data, can you provide labels for all of them, e.g. the 3rd, blackburn_et_al_2011:E?

@damianooldoni can you have a look into this ?

> source: I've detected 2838 different values for that. Is it really the purpose that the user selects one or more of these? Should we provide option to select all sources?

a suggestion, take the 10 largest checklists and pool the rest in others   
@timadriaens what do you think ?

> You can see that the observation date is missing in almost 50% of the cases. This would mean that all these records are excluded from the graphs that display "periode" or filter on "periode".

This issue is known and also one of the reasons to make this page ! we want the checklist people to complete their checklists and publish them.

**Comment by damianooldoni** [comment]
@mvarewyck: `trias` package is here: https://github.com/trias-project/trias. You can install it via `devtools::install_github("trias-project/trias")`.

**Comment by damianooldoni** [comment]
About degree of establishment: I will update the input data @mvarewyck is using based on the mapping in table 2 at https://biss.pensoft.net/article/38084/element/3/4784909/. At TrIAS we will probably do the same while producing the unified checklist, but this could take longer. So, meanwhile we map it in this pipeline: https://trias-project.github.io/indicators/get_data_input_checklist_indicators.html

**Comment by damianooldoni** [comment]
I mapped the degree of establishment as asked in this issue.
See https://trias-project.github.io/indicators/get_data_input_checklist_indicators.html#231_mapping_degree_of_establishment for details about the mapping.
So, file https://github.com/trias-project/indicators/blob/master/data/interim/data_input_checklist_indicators.tsv contains the right changes.
I have also removed spaces in column names (`degree_of_establishment` instead of `degree of establishment`, ...). Notice also that `locationId` has been changed in unified checklist: `ISO_3166:BE` instead of `ISO_3166-2:BE`, so in this file.

**Comment by eadriaensen** [comment]
@SanderDevisscher @timadriaens 
>>source: I've detected 2838 different values for that. Is it really the purpose that the user selects one or more of these? Should we provide option to select all sources?

>a suggestion, take the 10 largest checklists (most unique taxons) and pool the rest in others
@timadriaens what do you think ?

Should I go ahead with the proposed suggestion of @SanderDevisscher concerning 'source'?
(Assuming that "largest checklist" means 'most frequently occuring'):


`head(sort(table(exotenData$source), decreasing = TRUE), n = 10)`

```
                                                       https://www.gbif.org/species/141266506: Epilobium ciliatum Rafin. in Verloove F, Groom Q, Brosens D, Desmet P, Reyserhove L (2019). Manual of the Alien Plants of Belgium. Version 1.8. Botanic Garden Meise. Checklist dataset https://doi.org/10.15468/wtda1m 
                                                                                                                                                                                                                                                                                                                    60 
                                                           https://www.gbif.org/species/141265920: Paspalum distichum L. in Verloove F, Groom Q, Brosens D, Desmet P, Reyserhove L (2019). Manual of the Alien Plants of Belgium. Version 1.8. Botanic Garden Meise. Checklist dataset https://doi.org/10.15468/wtda1m 
                                                                                                                                                                                                                                                                                                                    50 
                                             https://www.gbif.org/species/141264976: Ailanthus altissima (Mill.) Swingle in Verloove F, Groom Q, Brosens D, Desmet P, Reyserhove L (2019). Manual of the Alien Plants of Belgium. Version 1.8. Botanic Garden Meise. Checklist dataset https://doi.org/10.15468/wtda1m 
                                                                                                                                                                                                                                                                                                                    48 
                                                            https://www.gbif.org/species/141265977: Bunias orientalis L. in Verloove F, Groom Q, Brosens D, Desmet P, Reyserhove L (2019). Manual of the Alien Plants of Belgium. Version 1.8. Botanic Garden Meise. Checklist dataset https://doi.org/10.15468/wtda1m 
                                                                                                                                                                                                                                                                                                                    45 
                      https://www.gbif.org/species/159498548: Oxyura jamaicensis (Gmelin, 1789) in Preda C, Adriaens T, Strubbe D, Desmet P, Reyserhove L (2019). Checklist of alien birds of Belgium. Version 1.2. Research Institute for Nature and Forest (INBO). Checklist dataset https://doi.org/10.15468/wr3gis 
                                                                                                                                                                                                                                                                                                                    42 
                                                 https://www.gbif.org/species/141264827: Hydrocotyle ranunculoides L. f. in Verloove F, Groom Q, Brosens D, Desmet P, Reyserhove L (2019). Manual of the Alien Plants of Belgium. Version 1.8. Botanic Garden Meise. Checklist dataset https://doi.org/10.15468/wtda1m 
                                                                                                                                                                                                                                                                                                                    40 
https://www.gbif.org/species/141117247: Neogobius melanostomus in Verreycken H, Belpaire C, Adriaens T, Desmet P, Reyserhove L (2018). Checklist of non-native freshwater fishes in Flanders, Belgium. Version 1.4. Research Institute for Nature and Forest (INBO). Checklist dataset https://doi.org/10.15468/xvuzfh 
                                                                                                                                                                                                                                                                                                                    36 
                                                        https://www.gbif.org/species/141264935: Rhododendron ponticum L. in Verloove F, Groom Q, Brosens D, Desmet P, Reyserhove L (2019). Manual of the Alien Plants of Belgium. Version 1.8. Botanic Garden Meise. Checklist dataset https://doi.org/10.15468/wtda1m 
                                                                                                                                                                                                                                                                                                                    36 
                                               https://www.gbif.org/species/141265477: Cotoneaster horizontalis Decaisne in Verloove F, Groom Q, Brosens D, Desmet P, Reyserhove L (2019). Manual of the Alien Plants of Belgium. Version 1.8. Botanic Garden Meise. Checklist dataset https://doi.org/10.15468/wtda1m 
                                                                                                                                                                                                                                                                                                                    32 
                           https://www.gbif.org/species/159498554: Anser indicus (Latham, 1790) in Preda C, Adriaens T, Strubbe D, Desmet P, Reyserhove L (2019). Checklist of alien birds of Belgium. Version 1.2. Research Institute for Nature and Forest (INBO). Checklist dataset https://doi.org/10.15468/wr3gis 
                                                                                                                                                                                                                                                                                                                    32 

```

**Comment by SanderDevisscher** [comment]
@eadriaensen now it gets complicated (I think). The actual sources for the records are the "Manual of the Alien Plants of Belgium", "Checklist of alien birds of Belgium" and Checklist of non-native freshwater fishes in Flanders, Belgium. So the string between `(jjjj).` and version `x.x`. I don't know if this is regexable but the rest is filler.

**Comment by eadriaensen** [comment]
Okay, let me give that a go.
So with that I assume that the 10 largest checklists ( i.e.  `source`) should be selected in a data driven manner.
Should this be determined on the level of raw data  - `data_input_checklist_indicators.tsv`?
Or on the level of filtered data - i.e. re-evaluated while using the app depending on user choices?

**Comment by SanderDevisscher** [comment]
> So with that I assume that the 10 largest checklists ( i.e. source) should be selected in a data driven manner.

yes

> Should this be determined on the level of raw data - data_input_checklist_indicators.tsv?
> Or on the level of filtered data - i.e. re-evaluated while using the app depending on user choices?

The 10 largest checklists should be determined based on filtered data, so option 2.

**Comment by timadriaens** [comment]
The "sources" refer to the **checklists** published, which you can find via the [TrIAS network page on gbif](https://www.gbif.org/network/b153643d-735a-440f-a0e9-428b4f9d1cd2). So they include:

- Checklist of alien birds of Belgium : Alien Bird Checklist
- World Register of Introduced Marine Species (WRiMS): WRiMS
- Registry of introduced terrestrial molluscs in Belgium: Alien Mollusc checklist
- Manual of the Alien Plants of Belgium: Manual of Alien Plants
- Inventory of alien macroinvertebrates in Flanders, Belgium: Alien Macroinverts
- Checklist of non-native freshwater fishes in Flanders, Belgium: Alien Fish Flanders
- Catalogue of the Rust Fungi of Belgium: Belgian rust fungi
- RINSE - Registry of non-native species in the Two Seas region countries (Great Britain, France, Belgium and the Netherlands): RINSE1
- RINSE - Pathways and vectors of biological invasions in Northwest Europe: RINSE2
- Ad hoc checklist of alien species in Belgium: Add hoc alien species checklist

I have made a proposal of shortnames for those so that they can figure in a selection pane.

**Comment by eadriaensen** [comment]
Thanks for your replies, @SanderDevisscher @timadriaens 
After formatting the `source` variable, the following summary is obtained:
```
> table(exotenData$source, useNA = "ifany")

                            Ad hoc checklist of alien species in Belgium 
                                                                     548 
                                     Checklist of alien birds of Belgium 
                                                                     564 
          Checklist of non-native freshwater fishes in Flanders, Belgium 
                                                                     166 
              Inventory of alien macroinvertebrates in Flanders, Belgium 
                                                                     238 
                                   Manual of the Alien Plants of Belgium 
                                                                    8587 
RINSE - Pathways and vectors of biological invasions in Northwest Europe 
                                                                     120 
                     World Register of Introduced Marine Species (WRiMS) 
                                                                      41 
                                                                    <NA> 
                                                                     111 
```

I will further format with the suggested shortened names.
Note that of the listed sources by @timadriaens , the sources listed below are currently not present in the raw data:

> * Registry of introduced terrestrial molluscs in Belgium: Alien Mollusc checklist
> * Catalogue of the Rust Fungi of Belgium: Belgian rust fungi
> * RINSE - Registry of non-native species in the Two Seas region countries (Great Britain, France, Belgium and the Netherlands): RINSE1

**Comment by eadriaensen** [comment]
@SanderDevisscher @timadriaens 

Hoe zien jullie het keuze menu concreet?
Zit er een soort hierarchie in de selectie? vb. eerst locality en periode, dan de daarvan beschikbare sources, ... etc.
Het gaat hier om de keuzes ivm `periode`, `regio`, `degree of establishment`, `habitat`, `bron`, `kingdom` en `pathwayway_level1`
De hierarchie kingdom -> phylum -> class -> order -> family en  pathway_level 1 > pathway_level 2 is duidelijk.

Andere optie is om de 'algemene' keuzes eerst te laten maken en daarna te kijken of er data 'voldoen' aan de gemaakte keuzes en zo ja, welke keuzes er dan overblijven voor taxonomy en pathway

**Comment by SanderDevisscher** [comment]
Ik zou persoonlijk voor optie 2 gaan met name:
Periode en regio bepalen de default grafieken. De andere filters beginnen leeg en filteren de data dus niet. M.a.w.  zonder de filters aan te passen krijg je de grafieken en tabellen voor België tussen 1950 en nu, van alle taxa, degrees of establishment, habitats en bronnen. 
Wanneer je één van de filters aanpast worden de lijsten in de andere filters hierdoor aangepast. 
Voorbeeld1: je selecteert het Kingdom "Plantae", uit Belgische data van 1950 tot nu worden de Plantae gefilterd. De filter phylum wordt nu toegevoegd met enkel de phyla die voorkomen in de gefilterde dataset.  Hetzelfde gebeurd nu voor de andere filters, ze worden gevuld met de opties die in de gefilterde data voorkomen.  
Voorbeeld2: je selecteert de bron "manual of alien plants" en past de slider aan naar 1930 - 1950, Uit de Belgische data van 1930 t.e.m. 1950 worden alle taxa gefilterd met bron "manual of alien plants". Hierdoor wordt de Kingdom filter beperkt tot Plantae omdat dit het enige kingdom is in de data. Hetzelfde gebeurd nu voor de andere filters, ze worden gevuld met de opties die in de gefilterde data voorkomen.  

@eadriaensen hopelijk is dit verduidelijkend

**Comment by SanderDevisscher** [comment]
@eadriaensen @mvarewyck is it possible to add an additional filter for union list species? 
The default for this filter should be to include all species (on and off the list). 

The union list can be found here:  https://github.com/trias-project/indicators/blob/master/data/input/eu_concern_species.tsv

**Comment by eadriaensen** [comment]
@SanderDevisscher  Can you confirm that I understand the request right:

* An additional filter should be added that filters the dataset (checklist_indicators) on the column `species` (or `canonical name` ?).

There are more missings in `species` than in `canonical name`:

```
> sum(rawData$species %in% c(NA, " ", ""))
[1] 1028
> sum(rawData$canonicalName %in% c(NA, " ", ""))
[1] 15
```

* Possible choices for this filter should be any species present in the `species` column (checklist_indicators dataset), that is also specified in the `checklist_scientificName` column of the union list ? There are 416 such cases

```
> length(which(rawData$species %in% unionlist$checklist_scientificName))
[1] 416
```

* Without user choice (i.e. by default) the checklist_indicators is not filtered for the `species` column, i.e. all species are included irrespective of their presence in the union list?

**Comment by SanderDevisscher** [comment]
That's correct. Filtration should occur on the species level. however since the species level seems not to be complete the canonical name should be used when species is missing.

> Possible choices for this filter should be any species present in the species column (checklist_indicators dataset), that is also specified in the checklist_scientificName column of the union list ? There are 416 such cases

choices should be:
- all species
- only eu concern species

> Without user choice (i.e. by default) the checklist_indicators is not filtered for the species column, i.e. all species are included irrespective of their presence in the union list?

correct
